bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/ThingStatusTriggerHandler.java

```diff
@@ -0,0 +1,146 @@
+        this.status = (String) module.getConfiguration().get(CFG_STATUS);
+        this.previousStatus = (String) module.getConfiguration().get(CFG_PREVIOUS_STATUS);
+        if (UPDATE_MODULE_TYPE_ID.equals(module.getTypeUID())) {
+            this.types = Collections.singleton(ThingStatusInfoEvent.TYPE);
+        } else {
+            this.types = Collections.singleton(ThingStatusInfoChangedEvent.TYPE);
+        }
+        this.bundleContext = bundleContext;
+        Dictionary<String, Object> properties = new Hashtable<>();
+        properties.put("event.topics", "smarthome/things/" + thingUID + "/*");
+        eventSubscriberRegistration = this.bundleContext.registerService(EventSubscriber.class.getName(), this,
+                properties);
+    }
+
+    @Override
+    public Set<String> getSubscribedEventTypes() {
+        return types;
+    }
+
+    @Override
+    public @Nullable EventFilter getEventFilter() {
+        return this;
+    }
+
+    @Override
+    public void receive(Event event) {
+        if (callback != null) {
+            logger.trace("Received Event: Source: {} Topic: {} Type: {}  Payload: {}", event.getSource(),
+                    event.getTopic(), event.getType(), event.getPayload());
+            Map<String, Object> values = new HashMap<>();
+            if (event instanceof ThingStatusInfoEvent && UPDATE_MODULE_TYPE_ID.equals(module.getTypeUID())) {
+                ThingStatus status = ((ThingStatusInfoEvent) event).getStatusInfo().getStatus();
+                if (statusMatches(this.status, status)) {
+                    values.put(OUT_STATUS, status);
+                }
+            } else if (event instanceof ThingStatusInfoChangedEvent
+                    && CHANGE_MODULE_TYPE_ID.equals(module.getTypeUID())) {
+                ThingStatus newStatus = ((ThingStatusInfoChangedEvent) event).getStatusInfo().getStatus();
+                ThingStatus oldStatus = ((ThingStatusInfoChangedEvent) event).getOldStatusInfo().getStatus();
+                if (statusMatches(this.status, newStatus) && statusMatches(this.previousStatus, oldStatus)) {
+                    values.put(OUT_NEW_STATUS, newStatus);
+                    values.put(OUT_OLD_STATUS, oldStatus);
+                }
+            }
+            if (!values.isEmpty()) {
+                values.put(OUT_EVENT, event);
+                ((TriggerHandlerCallback) callback).triggered(module, values);
+            }
+        }
+    }
+
+    private boolean statusMatches(@Nullable String requiredStatus, ThingStatus status) {
+        if (requiredStatus == null) {
+            return true;
+        }
+        String reqState = requiredStatus.trim();
+        return reqState.isEmpty() || reqState.equals(status.toString());
+    }
+
+    /**
+     * do the cleanup: unregistering eventSubscriber...
+     */
+    @Override
+    public void dispose() {
+        super.dispose();
+        eventSubscriberRegistration.unregister();
```

Comment:
Also a very minor comment:

Shouldn't we do our "more specific" dispose logic apply in front of the dispose logic of the base class.
That's more a generic lifecycle handling (similar to constructor and destructor (if available) or activate, deactivate, ...)

1. init base
1. init specific
1. ...
1. deinit specific
1. deinit base